pkg/quota/controller/clusterquotareconciliation/reconcilation_controller.go

```diff
@@ -0,0 +1,356 @@
+
+	resyncPeriod time.Duration
+
+	// queue tracks which clusterquotas to update along with a list of namespaces for that clusterquota
+	queue BucketingWorkQueue
+
+	// knows how to calculate usage
+	registry utilquota.Registry
+	// controllers monitoring to notify for replenishment
+	replenishmentControllers []framework.ControllerInterface
+}
+
+type workItem struct {
+	namespaceName      string
+	forceRecalculation bool
+}
+
+func NewClusterQuotaReconcilationController(options ClusterQuotaReconcilationControllerOptions) *ClusterQuotaReconcilationController {
+	c := &ClusterQuotaReconcilationController{
+		clusterQuotaLister: options.ClusterQuotaInformer.Lister(),
+		clusterQuotaSynced: options.ClusterQuotaInformer.Informer().HasSynced,
+		clusterQuotaMapper: options.ClusterQuotaMapper,
+		clusterQuotaClient: options.ClusterQuotaClient,
+
+		resyncPeriod: options.ResyncPeriod,
+		registry:     options.Registry,
+
+		queue: NewBucketingWorkQueue(),
+	}
+
+	// we need to trigger every time
+	options.ClusterQuotaInformer.Informer().AddEventHandler(framework.ResourceEventHandlerFuncs{
+		AddFunc:    c.addClusterQuota,
+		UpdateFunc: c.updateClusterQuota,
+	})
+
+	for _, groupKindToReplenish := range options.GroupKindsToReplenish {
+		controllerOptions := &resourcequota.ReplenishmentControllerOptions{
+			GroupKind:         groupKindToReplenish,
+			ResyncPeriod:      options.ReplenishmentResyncPeriod,
+			ReplenishmentFunc: c.replenishQuota,
+		}
+		replenishmentController, err := options.ControllerFactory.NewController(controllerOptions)
+		if err != nil {
+			glog.Warningf("quota controller unable to replenish %s due to %v, changes only accounted during full resync", groupKindToReplenish, err)
+		} else {
+			c.replenishmentControllers = append(c.replenishmentControllers, replenishmentController)
+		}
+	}
+	return c
+}
+
+// Run begins quota controller using the specified number of workers
+func (c *ClusterQuotaReconcilationController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+
+	// Wait for the stores to sync before starting any work in this controller.
+	ready := make(chan struct{})
+	go c.waitForSyncedStores(ready, stopCh)
+	select {
+	case <-ready:
+	case <-stopCh:
+		return
+	}
+
+	// the controllers that replenish other resources to respond rapidly to state changes
+	for _, replenishmentController := range c.replenishmentControllers {
+		go replenishmentController.Run(stopCh)
+	}
+
+	// the workers that chug through the quota calculation backlog
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+
+	// the timer for how often we do a full recalculation across all quotas
+	go wait.Until(func() { c.calculateAll() }, c.resyncPeriod, stopCh)
+
+	<-stopCh
+	glog.Infof("Shutting down ClusterQuotaReconcilationController")
+	c.queue.ShutDown()
+}
+
+func (c *ClusterQuotaReconcilationController) waitForSyncedStores(ready chan<- struct{}, stopCh <-chan struct{}) {
```

Comment:
> if we're only watching a single object, why do we need to wait for a sync? is it because the same cache is backing the mapper?

This controller effectively watches namespaces as well, since its wire up as a cluster quota mapper listener.